[Heartbeat] Monitor Retries

andrewvc opened the pull request:
After experimenting around a bit, this gets retries working, using code at the monitor level. This also really neatly refactors the summary and state wrappers into much easier to understand code. This is an experiment only. Code is still pretty ugly and we should reorganize parameters and objects into something neater than what's here.

The main question this seeks to answer is __where__ we should put this logic.

Sample output of:

```
x-pack/heartbeat$  mage build && ELASTIC_SYNTHETICS_CAPABLE=true ./heartbeat -e | jq -c  '{"monitor": {"name": .monitor.name, "status": .monitor.status}, "summary": .summary, "state": .state}'
{"monitor":{"name":"Test","status":"up"},"summary":null,"state":null}
{"monitor":{"name":"Test","status":"up"},"summary":null,"state":null}
{"monitor":{"name":"Test","status":"up"},"summary":null,"state":null}
{"monitor":{"name":"Test","status":"up"},"summary":{"status":"up","retry_group":"R-netflix-1898aee1e84","attempt":1,"max_attempts":2,"final_attempt":false,"up":5,"down":0},"state":{"duration_ms":"0","up":1,"id":"default-1898aee1ec0-0","started_at":"2023-07-24T21:43:28.576999-05:00","status":"up","checks":1,"down":0,"flap_history":[],"ends":null}}
{"monitor":{"name":"Test","status":"up"},"summary":null,"state":null}
{"monitor":{"name":"Test","status":"up"},"summary":null,"state":null}
{"monitor":{"name":"Test","status":"up"},"summary":null,"state":null}
{"monitor":{"name":"Test","status":"up"},"summary":{"up":5,"down":0,"status":"up","retry_group":"R-netflix-1898aee1e84","attempt":2,"max_attempts":2,"final_attempt":true},"state":{"id":"default-1898aee1ec0-0","started_at":"2023-07-24T21:43:28.576999-05:00","duration_ms":"23","checks":2,"down":0,"ends":null,"status":"up","up":2,"flap_history":[]}}
```


## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- 

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
Pinging @elastic/uptime (Team:Uptime)

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-36147/detail/PR-36147/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-36147/detail/PR-36147/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-36147/detail/PR-36147/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-36147/detail/PR-36147/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2023-07-25T02:38:25.429Z&rangeTo=2023-07-25T02:58:25.429Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=3c19fe3319b19a4b78376acab67148b1&transactionId=f72d8895ffd9b204)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-07-25T02:48:25.429+0000


* Duration: 18 min 26 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 6  |
| Passed       | 156  |
| Skipped      | 0 |
| Total        | 162   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![6](https://img.shields.io/badge/6%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-36147/detail/PR-36147/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Build&Test / x-pack/heartbeat-unitTest / TestSimpleScenariosBasicFields/tcp-simple – github.com/elastic/beats/v7/x-pack/heartbeat/scenarios`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestSimpleScenariosBasicFields/tcp-simple
=== PAUSE TestSimpleScenariosBasicFields/tcp-simple
=== CONT  TestSimpleScenariosBasicFields/tcp-simple
=== CONT  TestSimpleScenariosBasicFields/tcp-simple
    testing.go:36: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/pkg/mod/github.com/elastic/go-lookslike@v0.3.0/testslike/testing.go:36
        	Error:      	lookslike could not validate map
        	Test:       	TestSimpleScenariosBasicFields/tcp-simple
        	Messages:   	1 errors validating source: 
        	            	(mapstr.M) (len=6) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	 (string) (len=3) "tcp": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=3) "rtt": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=7) "connect": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 1106
        	            	   }
        	            	  }
        	            	 },
        	            	 (string) (len=7) "monitor": (mapstr.M) (len=7) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=2) "id": (string) (len=11) "tcp-test-id",
        	            	  (string) (len=6) "status": (string) (len=2) "up",
        	            	  (string) (len=8) "duration": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=2) "us": (int64) 1258
        	            	  },
        	            	  (string) (len=2) "ip": (string) (len=9) "127.0.0.1",
        	            	  (string) (len=8) "timespan": (mapstr.M) (len=2) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=3) "gte": (time.Time) 2023-07-25 03:06:14.848983325 +0000 UTC m=+0.018256070,
        	            	   (string) (len=2) "lt": (time.Time) 2023-07-25 03:07:14.848983325 +0000 UTC m=+60.018256070
        	            	  },
        	            	  (string) (len=4) "name": (string) (len=13) "tcp-test-name",
        	            	  (string) (len=4) "type": (string) (len=3) "tcp"
        	            	 },
        	            	 (string) (len=3) "url": (mapstr.M) (len=4) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=4) "port": (uint16) 46615,
        	            	  (string) (len=4) "full": (string) (len=21) "tcp://127.0.0.1:46615",
        	            	  (string) (len=6) "scheme": (string) (len=3) "tcp",
        	            	  (string) (len=6) "domain": (string) (len=9) "127.0.0.1"
        	            	 },
        	            	 (string) (len=7) "summary": (*monitors.JobSummary)(0xc00075e050)({
        	            	  Attempt: (uint32) 1,
        	            	  MaxAttempts: (uint32) 2,
        	            	  FinalAttempt: (bool) true,
        	            	  Up: (uint32) 1,
        	            	  Down: (uint32) 0,
        	            	  Status: (monitorstate.StateStatus) (len=2) "up",
        	            	  RetryGroup: (string) (len=13) "R-1898b02f7bf",
        	            	  mtx: (*sync.Mutex)(0xc00075a040)({
        	            	   state: (int32) 0,
        	            	   sema: (uint32) 0
        	            	  }),
        	            	  contsOutstanding: (uint32) 0,
        	            	  rootJob: (jobs.Job) 0x1051380
        	            	 }),
        	            	 (string) (len=5) "state": (*monitorstate.State)(0xc0007601b0)(<monitorstate:id=default-1898b02f7c1-0,started=2023-07-25 03:06:14.849053561 +0000 UTC m=+0.018326295,up=1,down=0>),
        	            	 (string) (len=5) "event": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=7) "dataset": (string) (len=3) "tcp"
        	            	 }
        	            	}
=== CONT  TestSimpleScenariosBasicFields/tcp-simple
    testing.go:44: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/pkg/mod/github.com/elastic/go-lookslike@v0.3.0/testslike/testing.go:44
        	Error:      	Received unexpected error:
        	            	@Path 'monitor.check_group': expected this key to be present
        	Test:       	TestSimpleScenariosBasicFields/tcp-simple
    basics_test.go:39: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/src/github.com/elastic/beats/x-pack/heartbeat/scenarios/basics_test.go:39
        	            				/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/src/github.com/elastic/beats/x-pack/heartbeat/scenarios/framework/framework.go:127
        	Error:      	Received unexpected error:
        	            	key not found
        	Test:       	TestSimpleScenariosBasicFields/tcp-simple
    --- FAIL: TestSimpleScenariosBasicFields/tcp-simple (0.01s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/heartbeat-unitTest / TestSimpleScenariosBasicFields/http-simple – github.com/elastic/beats/v7/x-pack/heartbeat/scenarios`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestSimpleScenariosBasicFields/http-simple
=== PAUSE TestSimpleScenariosBasicFields/http-simple
=== CONT  TestSimpleScenariosBasicFields/http-simple
=== CONT  TestSimpleScenariosBasicFields/http-simple
    testing.go:36: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/pkg/mod/github.com/elastic/go-lookslike@v0.3.0/testslike/testing.go:36
        	Error:      	lookslike could not validate map
        	Test:       	TestSimpleScenariosBasicFields/http-simple
        	Messages:   	1 errors validating source: 
        	            	(mapstr.M) (len=7) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	 (string) (len=5) "event": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=7) "dataset": (string) (len=4) "http"
        	            	 },
        	            	 (string) (len=3) "tcp": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=3) "rtt": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=7) "connect": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 286
        	            	   }
        	            	  }
        	            	 },
        	            	 (string) (len=4) "http": (mapstr.M) (len=2) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=8) "response": (mapstr.M) (len=4) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=7) "headers": (mapstr.M) (len=3) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=12) "Content-Type": (string) (len=25) "text/plain; charset=utf-8",
        	            	    (string) (len=4) "Date": (string) (len=29) "Tue, 25 Jul 2023 03:06:14 GMT",
        	            	    (string) (len=14) "Content-Length": (string) (len=2) "13"
        	            	   },
        	            	   (string) (len=11) "status_code": (int) 200,
        	            	   (string) (len=4) "body": (mapstr.M) (len=2) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=4) "hash": (string) (len=64) "68e656b251e67e8358bef8483ab0d51c6619f3e7a1a9f0e75838d41ff368f728",
        	            	    (string) (len=5) "bytes": (int) 13
        	            	   },
        	            	   (string) (len=9) "mime_type": (string) (len=25) "text/plain; charset=utf-8"
        	            	  },
        	            	  (string) (len=3) "rtt": (mapstr.M) (len=5) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=13) "write_request": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 127
        	            	   },
        	            	   (string) (len=8) "validate": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 646
        	            	   },
        	            	   (string) (len=7) "content": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 250
        	            	   },
        	            	   (string) (len=5) "total": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 1012
        	            	   },
        	            	   (string) (len=15) "response_header": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 396
        	            	   }
        	            	  }
        	            	 },
        	            	 (string) (len=7) "monitor": (mapstr.M) (len=7) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=2) "id": (string) (len=12) "http-test-id",
        	            	  (string) (len=4) "name": (string) (len=14) "http-test-name",
        	            	  (string) (len=6) "status": (string) (len=2) "up",
        	            	  (string) (len=8) "duration": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=2) "us": (int64) 1231
        	            	  },
        	            	  (string) (len=2) "ip": (string) (len=9) "127.0.0.1",
        	            	  (string) (len=8) "timespan": (mapstr.M) (len=2) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=3) "gte": (time.Time) 2023-07-25 03:06:14.84869143 +0000 UTC m=+0.017964168,
        	            	   (string) (len=2) "lt": (time.Time) 2023-07-25 03:07:14.84869143 +0000 UTC m=+60.017964168
        	            	  },
        	            	  (string) (len=4) "type": (string) (len=4) "http"
        	            	 },
        	            	 (string) (len=3) "url": (mapstr.M) (len=4) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=6) "domain": (string) (len=9) "127.0.0.1",
        	            	  (string) (len=4) "port": (uint16) 46615,
        	            	  (string) (len=4) "full": (string) (len=22) "http://127.0.0.1:46615",
        	            	  (string) (len=6) "scheme": (string) (len=4) "http"
        	            	 },
        	            	 (string) (len=7) "summary": (*monitors.JobSummary)(0xc0001381e0)({
        	            	  Attempt: (uint32) 1,
        	            	  MaxAttempts: (uint32) 2,
        	            	  FinalAttempt: (bool) true,
        	            	  Up: (uint32) 1,
        	            	  Down: (uint32) 0,
        	            	  Status: (monitorstate.StateStatus) (len=2) "up",
        	            	  RetryGroup: (string) (len=13) "R-1898b02f7bf",
        	            	  mtx: (*sync.Mutex)(0xc00012c520)({
        	            	   state: (int32) 0,
        	            	   sema: (uint32) 0
        	            	  }),
        	            	  contsOutstanding: (uint32) 0,
        	            	  rootJob: (jobs.Job) 0x1051380
        	            	 }),
        	            	 (string) (len=5) "state": (*monitorstate.State)(0xc000135200)(<monitorstate:id=default-1898b02f7c0-0,started=2023-07-25 03:06:14.848755811 +0000 UTC m=+0.018028559,up=1,down=0>)
        	            	}
=== CONT  TestSimpleScenariosBasicFields/http-simple
    testing.go:44: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/pkg/mod/github.com/elastic/go-lookslike@v0.3.0/testslike/testing.go:44
        	Error:      	Received unexpected error:
        	            	@Path 'monitor.check_group': expected this key to be present
        	Test:       	TestSimpleScenariosBasicFields/http-simple
    basics_test.go:39: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/src/github.com/elastic/beats/x-pack/heartbeat/scenarios/basics_test.go:39
        	            				/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/src/github.com/elastic/beats/x-pack/heartbeat/scenarios/framework/framework.go:127
        	Error:      	Received unexpected error:
        	            	key not found
        	Test:       	TestSimpleScenariosBasicFields/http-simple
    --- FAIL: TestSimpleScenariosBasicFields/http-simple (0.01s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/heartbeat-unitTest / TestSimpleScenariosBasicFields – github.com/elastic/beats/v7/x-pack/heartbeat/scenarios`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestSimpleScenariosBasicFields
--- FAIL: TestSimpleScenariosBasicFields (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/heartbeat-unitTest / TestLightweightSummaries/tcp-simple – github.com/elastic/beats/v7/x-pack/heartbeat/scenarios`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestLightweightSummaries/tcp-simple
=== PAUSE TestLightweightSummaries/tcp-simple
=== CONT  TestLightweightSummaries/tcp-simple
    testing.go:36: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/pkg/mod/github.com/elastic/go-lookslike@v0.3.0/testslike/testing.go:36
        	Error:      	lookslike could not validate map
        	Test:       	TestLightweightSummaries/tcp-simple
        	Messages:   	2 errors validating source: 
        	            	(mapstr.M) (len=6) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	 (string) (len=7) "summary": (*monitors.JobSummary)(0xc000040370)({
        	            	  Attempt: (uint32) 1,
        	            	  MaxAttempts: (uint32) 2,
        	            	  FinalAttempt: (bool) true,
        	            	  Up: (uint32) 1,
        	            	  Down: (uint32) 0,
        	            	  Status: (monitorstate.StateStatus) (len=2) "up",
        	            	  RetryGroup: (string) (len=13) "R-1898b02f848",
        	            	  mtx: (*sync.Mutex)(0xc000530530)({
        	            	   state: (int32) 0,
        	            	   sema: (uint32) 0
        	            	  }),
        	            	  contsOutstanding: (uint32) 0,
        	            	  rootJob: (jobs.Job) 0x1051380
        	            	 }),
        	            	 (string) (len=5) "state": (*monitorstate.State)(0xc000536f30)(<monitorstate:id=default-1898b02f849-0,started=2023-07-25 03:06:14.9851952 +0000 UTC m=+0.154467944,up=1,down=0>),
        	            	 (string) (len=5) "event": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=7) "dataset": (string) (len=3) "tcp"
        	            	 },
        	            	 (string) (len=3) "tcp": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=3) "rtt": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=7) "connect": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 235
        	            	   }
        	            	  }
        	            	 },
        	            	 (string) (len=7) "monitor": (mapstr.M) (len=7) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=4) "name": (string) (len=13) "tcp-test-name",
        	            	  (string) (len=4) "type": (string) (len=3) "tcp",
        	            	  (string) (len=2) "id": (string) (len=11) "tcp-test-id",
        	            	  (string) (len=6) "status": (string) (len=2) "up",
        	            	  (string) (len=8) "duration": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=2) "us": (int64) 370
        	            	  },
        	            	  (string) (len=2) "ip": (string) (len=9) "127.0.0.1",
        	            	  (string) (len=8) "timespan": (mapstr.M) (len=2) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=3) "gte": (time.Time) 2023-07-25 03:06:14.985166173 +0000 UTC m=+0.154438912,
        	            	   (string) (len=2) "lt": (time.Time) 2023-07-25 03:07:14.985166173 +0000 UTC m=+60.154438912
        	            	  }
        	            	 },
        	            	 (string) (len=3) "url": (mapstr.M) (len=4) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=6) "scheme": (string) (len=3) "tcp",
        	            	  (string) (len=6) "domain": (string) (len=9) "127.0.0.1",
        	            	  (string) (len=4) "port": (uint16) 46615,
        	            	  (string) (len=4) "full": (string) (len=21) "tcp://127.0.0.1:46615"
        	            	 }
        	            	}
    testing.go:44: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/pkg/mod/github.com/elastic/go-lookslike@v0.3.0/testslike/testing.go:44
        	Error:      	Received unexpected error:
        	            	@Path 'summary.down': expected this key to be present
        	Test:       	TestLightweightSummaries/tcp-simple
    testing.go:44: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/pkg/mod/github.com/elastic/go-lookslike@v0.3.0/testslike/testing.go:44
        	Error:      	Received unexpected error:
        	            	@Path 'summary.up': expected this key to be present
        	Test:       	TestLightweightSummaries/tcp-simple
    --- FAIL: TestLightweightSummaries/tcp-simple (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/heartbeat-unitTest / TestLightweightSummaries/http-simple – github.com/elastic/beats/v7/x-pack/heartbeat/scenarios`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestLightweightSummaries/http-simple
=== PAUSE TestLightweightSummaries/http-simple
=== CONT  TestLightweightSummaries/http-simple
=== CONT  TestLightweightSummaries/http-simple
    testing.go:36: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/pkg/mod/github.com/elastic/go-lookslike@v0.3.0/testslike/testing.go:36
        	Error:      	lookslike could not validate map
        	Test:       	TestLightweightSummaries/http-simple
        	Messages:   	2 errors validating source: 
        	            	(mapstr.M) (len=7) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	 (string) (len=4) "http": (mapstr.M) (len=2) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=8) "response": (mapstr.M) (len=4) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=7) "headers": (mapstr.M) (len=3) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=14) "Content-Length": (string) (len=2) "13",
        	            	    (string) (len=12) "Content-Type": (string) (len=25) "text/plain; charset=utf-8",
        	            	    (string) (len=4) "Date": (string) (len=29) "Tue, 25 Jul 2023 03:06:14 GMT"
        	            	   },
        	            	   (string) (len=11) "status_code": (int) 200,
        	            	   (string) (len=4) "body": (mapstr.M) (len=2) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=4) "hash": (string) (len=64) "68e656b251e67e8358bef8483ab0d51c6619f3e7a1a9f0e75838d41ff368f728",
        	            	    (string) (len=5) "bytes": (int) 13
        	            	   },
        	            	   (string) (len=9) "mime_type": (string) (len=25) "text/plain; charset=utf-8"
        	            	  },
        	            	  (string) (len=3) "rtt": (mapstr.M) (len=5) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=5) "total": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 810
        	            	   },
        	            	   (string) (len=13) "write_request": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 74
        	            	   },
        	            	   (string) (len=15) "response_header": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 333
        	            	   },
        	            	   (string) (len=8) "validate": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 537
        	            	   },
        	            	   (string) (len=7) "content": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 204
        	            	   }
        	            	  }
        	            	 },
        	            	 (string) (len=7) "monitor": (mapstr.M) (len=7) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=4) "type": (string) (len=4) "http",
        	            	  (string) (len=2) "id": (string) (len=12) "http-test-id",
        	            	  (string) (len=4) "name": (string) (len=14) "http-test-name",
        	            	  (string) (len=6) "status": (string) (len=2) "up",
        	            	  (string) (len=8) "duration": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=2) "us": (int64) 944
        	            	  },
        	            	  (string) (len=2) "ip": (string) (len=9) "127.0.0.1",
        	            	  (string) (len=8) "timespan": (mapstr.M) (len=2) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=3) "gte": (time.Time) 2023-07-25 03:06:14.98591365 +0000 UTC m=+0.155186394,
        	            	   (string) (len=2) "lt": (time.Time) 2023-07-25 03:07:14.98591365 +0000 UTC m=+60.155186394
        	            	  }
        	            	 },
        	            	 (string) (len=3) "url": (mapstr.M) (len=4) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=6) "scheme": (string) (len=4) "http",
        	            	  (string) (len=6) "domain": (string) (len=9) "127.0.0.1",
        	            	  (string) (len=4) "port": (uint16) 46615,
        	            	  (string) (len=4) "full": (string) (len=22) "http://127.0.0.1:46615"
        	            	 },
        	            	 (string) (len=7) "summary": (*monitors.JobSummary)(0xc0004b2960)({
        	            	  Attempt: (uint32) 1,
        	            	  MaxAttempts: (uint32) 2,
        	            	  FinalAttempt: (bool) true,
        	            	  Up: (uint32) 1,
        	            	  Down: (uint32) 0,
        	            	  Status: (monitorstate.StateStatus) (len=2) "up",
        	            	  RetryGroup: (string) (len=13) "R-1898b02f848",
        	            	  mtx: (*sync.Mutex)(0xc00012c030)({
        	            	   state: (int32) 0,
        	            	   sema: (uint32) 0
        	            	  }),
        	            	  contsOutstanding: (uint32) 0,
        	            	  rootJob: (jobs.Job) 0x1051380
        	            	 }),
        	            	 (string) (len=5) "state": (*monitorstate.State)(0xc0004474d0)(<monitorstate:id=default-1898b02f849-0,started=2023-07-25 03:06:14.985935971 +0000 UTC m=+0.155208717,up=1,down=0>),
        	            	 (string) (len=5) "event": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=7) "dataset": (string) (len=4) "http"
        	            	 },
        	            	 (string) (len=3) "tcp": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	  (string) (len=3) "rtt": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	   (string) (len=7) "connect": (mapstr.M) (len=1) (PANIC=runtime error: invalid memory address or nil pointer dereference){
        	            	    (string) (len=2) "us": (int64) 236
        	            	   }
        	            	  }
        	            	 }
        	            	}
    testing.go:44: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/pkg/mod/github.com/elastic/go-lookslike@v0.3.0/testslike/testing.go:44
        	Error:      	Received unexpected error:
        	            	@Path 'summary.up': expected this key to be present
        	Test:       	TestLightweightSummaries/http-simple
    testing.go:44: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-36147-1-f44bf2b1-92a8-4c78-afa4-bbbaf008194f/pkg/mod/github.com/elastic/go-lookslike@v0.3.0/testslike/testing.go:44
        	Error:      	Received unexpected error:
        	            	@Path 'summary.down': expected this key to be present
        	Test:       	TestLightweightSummaries/http-simple
    --- FAIL: TestLightweightSummaries/http-simple (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/heartbeat-unitTest / TestLightweightSummaries – github.com/elastic/beats/v7/x-pack/heartbeat/scenarios`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestLightweightSummaries
--- FAIL: TestLightweightSummaries (0.00s)
 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![4](https://img.shields.io/badge/4%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-36147/detail/PR-36147/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `x-pack/heartbeat-unitTest - mage build unitTest`
<ul>
<li>Took 3 min 16 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-36147/runs/1/steps/3105/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/heartbeat-unitTest - mage build unitTest`
<ul>
<li>Took 0 min 18 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-36147/runs/1/steps/3230/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/heartbeat-unitTest - mage build unitTest`
<ul>
<li>Took 0 min 18 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-36147/runs/1/steps/3234/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-36147/runs/1/steps/2736/log/?start=0">here</a></li>
<li>Description: <code>Error 'org.jenkinsci.plugins.workflow.steps.FlowInterruptedException'</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

emilioalvap commented on heartbeat/monitors/task.go:
```diff
@@ -103,18 +106,98 @@ func (t *configuredJob) Stop() {
+			logp.L().Errorf("monitor status found, but wasn't a string: %v", ms)
+		}
+
+		if msStr == "up" {
```
nit
```suggestion
		if msStr == monitorstate.StatusUp {
```

emilioalvap commented on heartbeat/monitors/wrappers/wrappers.go:
```diff
@@ -64,55 +63,136 @@ func WrapLightweight(js []jobs.Job, stdMonFields stdfields.StdMonitorFields, mst
-			}
+func addSummarizer(sf stdfields.StdMonitorFields, mst *monitorstate.Tracker, maxAttempts uint16) jobs.JobWrapper {
+	return jobs.WrapStateful[*Summarizer](func(rootJob jobs.Job) jobs.StatefulWrapper[*Summarizer] {
+		return newSummarizer(rootJob, sf, mst, maxAttempts)
```
q, is there a motivation to propagate `maxAttempts` all the way here?

emilioalvap commented on heartbeat/monitors/wrappers/wrappers.go:
```diff
@@ -64,55 +63,136 @@ func WrapLightweight(js []jobs.Job, stdMonFields stdfields.StdMonitorFields, mst
-
+	var swrapped []jobs.Job
+	for _, wj := range wrapped {
+		swrapped = append(swrapped, addSummarizer(stdMonFields, mst, maxAttempts)(wj))
```
Might be a `nit:`, but `StatefulFactory` overlaps significantly in intention with `jobs.WrapSeparately`. It's possible to get the same output by slightly editing JobWrapperFactory interface(unused atm) without these additional types:
```
wrapped := jobs.WrapAllSeparately(
    jobs.WrapAll(...),
    func(j jobs.Job) jobs.JobWrapper {
	s := newSummarizer(j, stdMonFields, mst, maxAttempts)
	return wrapSummarizer(s)
    },
```
If possible, I'd like to avoid introducing a new category of wrapper mechanism, so we could pick either one

andrewvc replied:
This is addressed in the latest version of this

vigneshshanmugam commented on heartbeat/monitors/stdfields/stdfields.go:
```diff
@@ -51,10 +52,13 @@ type StdMonitorFields struct {
 
 func ConfigToStdMonitorFields(conf *config.C) (StdMonitorFields, error) {
-	sFields := StdMonitorFields{Enabled: true}
+	sFields := StdMonitorFields{Enabled: true, MaxAttempts: 1}
```
I feel like we should make this configurable through env variable so we can control it on the service in a easier way than modifying the HB configs? Thoughts?

andrewvc replied:
This is a requirement for now, I'm glad to do it in a follow-up PR though! Let's see if there's a way we can do it that won't break the current Kibana UI. By requiring the config option only newer kibanas will see the new schema.

andrewvc commented on x-pack/heartbeat/scenarios/framework/fakeloader.go:
```diff
@@ -27,29 +29,33 @@ func newLoaderDB() *loaderDB {
 }
 
-func (ldb loaderDB) AddState(sf stdfields.StdMonitorFields, state *monitorstate.State) {
+func (ldb *loaderDB) AddState(sf stdfields.StdMonitorFields, state *monitorstate.State) {
```
I don't think the pointer matters for us, but it maxes the mutation more intuitive to callers.

andrewvc commented on x-pack/heartbeat/scenarios/framework/framework.go:
```diff
@@ -40,24 +40,32 @@ type Scenario struct {
-	}
+func MultiTwist(twists ...*Twist) *Twist {
+	return MakeTwist(
+		"<~MULTI-TWIST~[",
```
Makes things more legible in the console

ycombinator replied:
I'd suggest moving this comment into the code for future developers' convenience.

 ```suggestion
		"<~MULTI-TWIST~[", // Makes things more legible in the console
```

andrewvc commented on x-pack/heartbeat/scenarios/framework/fakeloader.go:
```diff
@@ -27,29 +29,33 @@ func newLoaderDB() *loaderDB {
-	ldb.lastTime = time.Now()
-
-	ldb.keysToState[monitorstate.LoaderDBKey(sf, ldb.lastTime, 0)] = state
+	key := keyFor(sf)
```
The previous behavior here was broken, but in a way that didn't matter for previous tests. Basically, the previous state would not be read unless the timestamps matched, which worked when all data was unix epoch. That's not the actual behavior of a simulated ES. We should simply load the last state. This was a misuse of the `LoaderDBKey` which was only ever meant to be used to append data, not to deterministically query old data.

andrewvc commented on x-pack/heartbeat/monitors/browser/synthexec/enrich_test.go:
```diff
@@ -125,29 +123,9 @@ func TestJourneyEnricher(t *testing.T) {
-				check(t, se, je)
-			}
-		})
+	je := makeTestJourneyEnricher(sFields)
```
Since we don't actually run legacy browsers anymore we can simplify here

andrewvc commented on x-pack/heartbeat/scenarios/stateloader_test.go:
```diff
@@ -16,14 +16,14 @@ import (
 		lastSS := framework.LastState(mtr.Events())
 
-		require.Equal(t, monitorstate.StatusUp, lastSS.State.Status)
+		assert.Equal(t, monitorstate.StatusUp, lastSS.State.Status)
```
Asserts here make understanding what failed easier. Otherwise you just get the length being wrong.

andrewvc commented on heartbeat/monitors/wrappers/monitorstate/monitorstate.go:
```diff
@@ -124,14 +126,16 @@ func (s *State) recordCheck(sf stdfields.StdMonitorFields, newStatus StateStatus
 		}
 
-		if !hasStabilized { // continue flapping
+		if !hasStabilized || !isFinalAttempt { // continue flapping
```
We still don't have flapping turned on, but I thought I'd at least adapt it to work right here

emilioalvap commented on heartbeat/monitors/wrappers/wrappers.go:
```diff
@@ -36,42 +34,44 @@ import (
 	}
+	// Wrap just the root jobs with the summarizer
+	// The summarizer itself wraps the continuations in a stateful way
+	for i, j := range wrapped {
+		j := j
+		wrapped[i] = func(event *beat.Event) ([]jobs.Job, error) {
+			s := summarizer.NewSummarizer(j, stdMonFields, mst)
+			return s.Wrap(j)(event)
+		}
+	}
+	return wrapped
```
```suggestion
	return jobs.WrapAllSeparately(
		wrapped,
		func(j jobs.Job) jobs.JobWrapper {
			s := summarizer.NewSummarizer(j, stdMonFields, mst)
			return s.Wrap
		},
	)
```
nit: for consistency. Doing these from GH UI, so there might be some syntax errors, sorry about that.

andrewvc replied:
I've just pushed a new commit that actually removes `WrapAllSeparately` as well as some other junk this PR accumulated in `job.go`. It wasn't actually used anywhere, and I maintain that it's simpler having the logic inline IMHO.

emilioalvap reviewed (commented):
Still reviewing logic changes, a couple of nits around syntax consistency. Edit: I think https://github.com/elastic/beats/issues/35465 should additionally be closed through this PR, I think all remaining logic was cleaned

emilioalvap commented on heartbeat/monitors/monitor.go:
```diff
@@ -178,7 +181,8 @@ func newMonitorUnsafe(
 		// effectively a lightweight job
-		wrappedJobs = wrappers.WrapLightweight(p.Jobs, m.stdFields, monitorstate.NewTracker(stateLoader, false))
+		m.stdFields.BadConfig = true
+		wrappedJobs = wrappers.WrapCommon(p.Jobs, m.stdFields, stateLoader)
```
Small optimization, `max_attempts` should be overridden in case of bad config, or include a check on the retry logic. Atm, its retrying monitors with bad config

emilioalvap commented on heartbeat/monitors/wrappers/summarizer/summarizer.go:
```diff
@@ -0,0 +1,152 @@
+
+type JobSummary struct {
+	Attempt      uint16                   `json:"attempt"`
+	MaxAttempts  uint16                   `json:"max_attempts"`
```
For any `max_attempts` value > 2, I only seem to get one retry on the UI. Tested with the following:
```yaml
- type: http
  enabled: true
  max_attempts: 5
  id: http_down_3
  name: http_down_3
  urls: ["http://localhost:9200"]
  schedule: '@every 1m'
```

emilioalvap commented on heartbeat/monitors/wrappers/summarizer/summarizer.go:
```diff
@@ -0,0 +1,152 @@
+				"state":   ms,
+			})
+
+			logp.L().Debugf("retry info: %v == %v && %d < %d", js.Status, lastStatus, js.Attempt, js.MaxAttempts)
```
```suggestion
			logp.L().Debugf("attempt info: %v == %v && %d < %d", js.Status, lastStatus, js.Attempt, js.MaxAttempts)
```
Found this log a bit confusing since first attempt is not technically a retry

emilioalvap commented:
@andrewvc I'm seeing some time resolution issues on the index on my tests due to lightweight checks being relatively fast. Three monitors with the same response pattern: `down` -> `up`-> `up`-> `up`-> `up`-> `down`

```
// Monitor 1
{"@timestamp":"2023-08-25T22:24:56.563+0200","message":"Monitor finished","monitor":{"id":"http_down_1","type":"http","status":"down","duration":{"ms":1}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:57.536+0200","message":"Monitor finished","monitor":{"id":"http_down_1","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:57.537+0200","message":"Monitor finished","monitor":{"id":"http_down_1","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:57.538+0200","message":"Monitor finished","monitor":{"id":"http_down_1","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:57.538+0200","message":"Monitor finished","monitor":{"id":"http_down_1","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:57.539+0200","message":"Monitor finished","monitor":{"id":"http_down_1","type":"http","status":"down","duration":{"ms":0}},"ecs.version":"1.6.0"}
// Monitor 2
{"@timestamp":"2023-08-25T22:24:56.563+0200","message":"Monitor finished","monitor":{"id":"http_down_2","type":"http","status":"down","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:56.882+0200","message":"Monitor finished","monitor":{"id":"http_down_2","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:56.883+0200","message":"Monitor finished","monitor":{"id":"http_down_2","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:56.884+0200","message":"Monitor finished","monitor":{"id":"http_down_2","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:56.884+0200","message":"Monitor finished","monitor":{"id":"http_down_2","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:56.885+0200","message":"Monitor finished","monitor":{"id":"http_down_2","type":"http","status":"down","duration":{"ms":0}},"ecs.version":"1.6.0"}
// Monitor 3
{"@timestamp":"2023-08-25T22:24:56.563+0200","message":"Monitor finished","monitor":{"id":"http_down_3","type":"http","status":"down","duration":{"ms":1}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:57.376+0200","message":"Monitor finished","monitor":{"id":"http_down_3","type":"http","status":"up","duration":{"ms":1}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:57.377+0200","message":"Monitor finished","monitor":{"id":"http_down_3","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:57.378+0200","message":"Monitor finished","monitor":{"id":"http_down_3","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:57.379+0200","message":"Monitor finished","monitor":{"id":"http_down_3","type":"http","status":"up","duration":{"ms":0}},"ecs.version":"1.6.0"}
{"@timestamp":"2023-08-25T22:24:57.379+0200","message":"Monitor finished","monitor":{"id":"http_down_3","type":"http","status":"down","duration":{"ms":0}},"ecs.version":"1.6.0"}
```
Producing divergent final statuses on the index:
![image](https://github.com/elastic/beats/assets/95703246/46d373f4-3b96-41f1-95a9-170d400e86da)
```
//Monitor 1 idx - desc time sort
"@timestamp","monitor.id","monitor.status"
"Aug 25, 2023 @ 22:24:57.538","http_down_1",up
"Aug 25, 2023 @ 22:24:57.538","http_down_1",down
"Aug 25, 2023 @ 22:24:57.537","http_down_1",up
"Aug 25, 2023 @ 22:24:57.537","http_down_1",up
"Aug 25, 2023 @ 22:24:57.536","http_down_1",up
"Aug 25, 2023 @ 22:24:56.562","http_down_1",down
//Monitor 2 idx - desc time sort
"@timestamp","monitor.id","monitor.status"
"Aug 25, 2023 @ 22:24:56.884","http_down_2",up
"Aug 25, 2023 @ 22:24:56.884","http_down_2",down
"Aug 25, 2023 @ 22:24:56.883","http_down_2",up
"Aug 25, 2023 @ 22:24:56.882","http_down_2",up
"Aug 25, 2023 @ 22:24:56.881","http_down_2",up
"Aug 25, 2023 @ 22:24:56.562","http_down_2",down
//Monitor 3 idx - desc time sort
"@timestamp","monitor.id","monitor.status"
"Aug 25, 2023 @ 22:24:57.379","http_down_3",down
"Aug 25, 2023 @ 22:24:57.378","http_down_3",up
"Aug 25, 2023 @ 22:24:57.377","http_down_3",up
"Aug 25, 2023 @ 22:24:57.376","http_down_3",up
"Aug 25, 2023 @ 22:24:57.375","http_down_3",up
"Aug 25, 2023 @ 22:24:56.562","http_down_3",down
```

I'd be happy to go through these.

emilioalvap commented on heartbeat/monitors/wrappers/summarizer/summarizer.go:
```diff
@@ -0,0 +1,152 @@
+		// Wrap downstream jobs using the same state object this lets us create new state
+		// on the first job, but re-use that same object on continuations.
+		for i, cont := range conts {
+			conts[i] = s.Wrap(cont)
```
@andrewvc I've noticed that for `mode: all` monitors with multiple resolved IPs, jobs will be retried `max_attempts` per resolved IP, not at monitor level. Might be worth mentioning that specifically somewhere

andrewvc commented:
@emilioalvap I've pushed a new summarizer test in https://github.com/elastic/beats/pull/36147/commits/e6d17857e0981372757d4d9279dfeec59189ee20 that confirms the results, but it broke a wrappers test by changing a common helper (execJobAndConts) that had weird behavior. I'll investigate more tomorrow.

emilioalvap commented on heartbeat/monitors/wrappers/summarizer/summarizer.go:
```diff
@@ -0,0 +1,173 @@
+				// We preserve `s` across attempts
+				s.jobSummary = NewJobSummary(js.Attempt+1, js.MaxAttempts, js.RetryGroup)
+				s.contsRemaining = 1
+				s.checkGroup = fmt.Sprintf("%s-%d", s.checkGroup, s.jobSummary.Attempt)
```
I've noticed this will generate the following sequence:
`1 st: uuid() -> 2nd uuid()+"1" -> 2nd uuid()+"-1" +"-2 -> nth uuid()+..."-<n-1>" +"-<n>"`

Is that the expected format?

andrewvc replied:
Yes, my thinking is that the UUID is already unique enough, and that it makes debugging easier to have the sequences. Also, you could potentially use a prefix query as an optimization in the future

emilioalvap commented on heartbeat/monitors/wrappers/summarizer/summarizer.go:
```diff
@@ -0,0 +1,173 @@
+	}
+}
+
+var id int
```
Doesn't seem like it's used effectively? If it's needed, it should be atomic uint imo.

emilioalvap commented on heartbeat/monitors/wrappers/summarizer/summarizer_test.go:
```diff
@@ -0,0 +1,150 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestSummarizer(t *testing.T) {
```
A couple other invariants that would be fairly easy to test: retry group remains equal across retries and check group is updated to reflect current attempt

andrewvc commented:
@emilioalvap thanks for the review feedback, I think I've addressed your concerns. Now if I can make the linter happy 🤞 I think the build may go green soon

emilioalvap reviewed (approved):
LGTM. One last concern, changes to `enrich.go` we discussed the other day about emitting summaries on `cmd/status` only don't seem to have been included yet: https://github.com/elastic/beats/blob/eef9598ea110abb7b4f37215fb396b0b32bdb2d8/x-pack/heartbeat/monitors/browser/synthexec/enrich.go#L144

ycombinator commented on heartbeat/monitors/wrappers/summarizer/summarizer_test.go:
```diff
@@ -0,0 +1,181 @@
+					lastSummary = summary
+				}
+				i += len(events)
+				if i >= len(tt.statusSequence) {
+					break
+				}
```
Nit: instead of this, you could change the `for` loop to:

```
			for i := 0; i < len(tt.statusSequence); {
```

ycombinator commented on heartbeat/tracer/tracer_test.go:
```diff
@@ -123,7 +123,8 @@ func listenTilClosed(t *testing.T, sockPath string) []string {
 	require.NoError(t, err)
-	var received []string
+	// no need to pre-allocate, but it seems to make the linter happy
+	received := make([]string, 0, 10)
```
Nit: you could omit the capacity here.
```suggestion
	received := make([]string, 0)
```

andrewvc replied:
Sure, but IIRC the linter wanted it in there, and it doesn't hurt.

ycombinator replied:
I think the linter was complaining more about the `nil` slice initialization.

ycombinator commented on x-pack/heartbeat/scenarios/testws.go:
```diff
@@ -0,0 +1,90 @@
+	testWsOnce.Do(func() {
+		testWs = httptest.NewServer(hbtest.HelloWorldHandler(200))
+
+		// wait for ws to become available
+		var err error
+		for i := 0; i < 20; i++ {
+			var resp *http.Response
+			req, _ := http.NewRequestWithContext(context.Background(), http.MethodGet, testWs.URL, nil)
+			resp, err = http.DefaultClient.Do(req)
+			if err == nil {
+				resp.Body.Close()
+				if resp.StatusCode == 200 {
+					break
+				}
+			}
+
+			time.Sleep(time.Millisecond * 250)
+		}
```
You could use `require.Eventually` here instead.
```suggestion
		// wait for ws to become available
		require.Eventually(t, func() bool {
			req, _ := http.NewRequestWithContext(context.Background(), http.MethodGet, testWs.URL, nil)
			resp, err := http.DefaultClient.Do(req)
			if err == nil {
				resp.Body.Close()
				return resp.StatusCode == 200
			}

			return false
		}, 5*time.Second, 250*time.Millisecond)
```

ycombinator commented on x-pack/heartbeat/scenarios/testws.go:
```diff
@@ -0,0 +1,90 @@
+		testWs = httptest.NewServer(hbtest.HelloWorldHandler(200))
+
+		// wait for ws to become available
+		var err error
+		for i := 0; i < 20; i++ {
+			var resp *http.Response
+			req, _ := http.NewRequestWithContext(context.Background(), http.MethodGet, testWs.URL, nil)
+			resp, err = http.DefaultClient.Do(req)
+			if err == nil {
+				resp.Body.Close()
+				if resp.StatusCode == 200 {
+					break
+				}
+			}
+
+			time.Sleep(time.Millisecond * 250)
+		}
+
+		if err != nil {
+			require.NoError(t, err, "could not retrieve successful response from test webserver")
+		}
```
You could use `require.Eventuallyf` here instead.
```suggestion
		require.Eventuallyf(t, func() bool {
			req, _ := http.NewRequestWithContext(context.Background(), http.MethodGet, testWs.URL, nil)
			resp, err := http.DefaultClient.Do(req)
			if err == nil {
				resp.Body.Close()
				return resp.StatusCode == 200
			}

			return false
		}, 5*time.Second, 250*time.Millisecond, "could not retrieve successful response from test webserver")
```

ycombinator commented on x-pack/heartbeat/scenarios/testws.go:
```diff
@@ -0,0 +1,90 @@
+	}))
+
+	// wait for ws to become available
+	var err error
+	for i := 0; i < 20; i++ {
+		var resp *http.Response
+		req, _ := http.NewRequestWithContext(context.Background(), http.MethodGet, testWs.URL, nil)
+		resp, err = http.DefaultClient.Do(req)
+		if err == nil {
+			resp.Body.Close()
+			if resp.StatusCode == 200 {
+				break
+			}
+		}
+
+		time.Sleep(time.Millisecond * 250)
+	}
+
+	if err != nil {
+		require.NoError(t, err, "could not retrieve successful response from test webserver")
+	}
```
See other comment about using `require.Eventuallyf` instead.

andrewvc replied:
fixed

ycombinator commented on x-pack/heartbeat/scenarios/framework/fakeloader.go:
```diff
@@ -27,29 +26,33 @@ func newLoaderDB() *loaderDB {
 
-func (ldb loaderDB) GetState(sf stdfields.StdMonitorFields) *monitorstate.State {
+func (ldb *loaderDB) GetState(sf stdfields.StdMonitorFields) *monitorstate.State {
 	ldb.mtx.Lock()
 	defer ldb.mtx.Unlock()
```
I know this change isn't part of this PR, but you could make `loaderDb.mtx` a `sync.RWMutex` and use `ldb.mtx.RLock()` and `ldb.mtx.RUnlock()` here.

andrewvc replied:
It's a good point, but this isn't performance critical code since it's just for testing.

ycombinator commented on x-pack/heartbeat/scenarios/basics_test.go:
```diff
@@ -34,20 +42,45 @@ func TestSimpleScenariosBasicFields(t *testing.T) {
+			var summary *summarizer.JobSummary
+			summaryIface, err := e.GetValue("summary")
+			if err == nil {
+				summary = summaryIface.(*summarizer.JobSummary)
```
See other comment about checking that the type assertion succeeded.

andrewvc replied:
See my comment on the previous comment

ycombinator reviewed (commented):
I'm far from an expert in Heartbeat / Synthetics so I didn't review the logic in this PR. But I made a few general Golang code review comments.

andrewvc commented on x-pack/heartbeat/scenarios/basics_test.go:
```diff
@@ -34,20 +42,45 @@ func TestSimpleScenariosBasicFields(t *testing.T) {
-				lastCg = cgStr
-			} else {
-				require.Equal(t, lastCg, cgStr)
+			cg := cgIface.(string)
```
I tend to skip that in tests because the verbosity is unnecessary, the worst case is it panics and we get a test failure. It's a little uglier, but I think it makes reading the test significantly easier by not cluttering it up with assertions that aren't getting at the thing we're actually testing.

andrewvc commented:
@ycombinator thanks for the review, I think I've addressed all of your comments

ycombinator reviewed (approved):
LGTM.

vigneshshanmugam commented on heartbeat/monitors/wrappers/summarizer/summarizer.go:
```diff
@@ -91,7 +91,7 @@ func (s *Summarizer) Wrap(j jobs.Job) jobs.Job {
 		conts, jobErr := j(event)
 
-		_, _ = event.PutValue("monitor.check_group", s.checkGroup)
+		_, _ = event.PutValue("monitor.check_group", fmt.Sprintf("%s-%d", s.checkGroup, s.jobSummary.Attempt))
```
@andrewvc Sorry for the delayed note on this PR, Would this break the existing uptime UI in any way.

andrewvc replied:
No, the UI treats the UUID as opaque values, as does ES (it's a keyword type, there's no dedicated UUID type).
